File: internal/apps/geo/gpx/web/gcj02.js
// GCJ-02, the system base maps published in China draw in, for positions the
// page holds itself — a planned route's waypoints — rather than receives drawn.
// The same conversion as internal/geo/gcj02, for [lon, lat] pairs.

const A = 6378245.0;
const EE = 0.00669342162296594323;

const inChina = ([lon, lat]) => lon >= 72.004 && lon <= 137.8347 && lat >= 0.8293 && lat <= 55.8271;

function transformLat(x, y) {
  let r = -100 + 2 * x + 3 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * Math.sqrt(Math.abs(x));
  r += ((20 * Math.sin(6 * x * Math.PI) + 20 * Math.sin(2 * x * Math.PI)) * 2) / 3;
  r += ((20 * Math.sin(y * Math.PI) + 40 * Math.sin((y / 3) * Math.PI)) * 2) / 3;
  r += ((160 * Math.sin((y / 12) * Math.PI) + 320 * Math.sin((y * Math.PI) / 30)) * 2) / 3;
  return r;
}

function transformLon(x, y) {
  let r = 300 + x + 2 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * Math.sqrt(Math.abs(x));
  r += ((20 * Math.sin(6 * x * Math.PI) + 20 * Math.sin(2 * x * Math.PI)) * 2) / 3;
  r += ((20 * Math.sin(x * Math.PI) + 40 * Math.sin((x / 3) * Math.PI)) * 2) / 3;
  r += ((150 * Math.sin((x / 12) * Math.PI) + 300 * Math.sin((x / 30) * Math.PI)) * 2) / 3;
  return r;
}

// fromWGS84 is the GCJ-02 position of a WGS-84 one.
export function fromWGS84(point) {
  if (!inChina(point)) return point;
  const [lon, lat] = point;
  const x = lon - 105, y = lat - 35;
  const radLat = (lat / 180) * Math.PI;
  let magic = Math.sin(radLat);
  magic = 1 - EE * magic * magic;
  const sqrtMagic = Math.sqrt(magic);
  const dLat = (transformLat(x, y) * 180) / (((A * (1 - EE)) / (magic * sqrtMagic)) * Math.PI);
  const dLon = (transformLon(x, y) * 180) / ((A / sqrtMagic) * Math.cos(radLat) * Math.PI);
  return [lon + dLon, lat + dLat];
}

// toWGS84 inverts fromWGS84, for a place clicked on a GCJ-02 map.
export function toWGS84(point) {
  if (!inChina(point)) return point;
  let [lon, lat] = point;
  for (let i = 0; i < 10; i++) {
    const [gLon, gLat] = fromWGS84([lon, lat]);
    lon -= gLon - point[0];
    lat -= gLat - point[1];
  }
  return [lon, lat];
}

File: internal/geo/gcj02/gcj02.go
// Package gcj02 converts WGS-84 positions to GCJ-02, the obfuscated system
// base maps published in China use (Gaode, Tencent). A WGS-84 track drawn on
// such a map is off by up to several hundred metres until converted.
//
// The conversion is the widely published reverse-engineered one, accurate to
// about 1–2 m. It is for display only: tracks keep their WGS-84 positions.
package gcj02

import (
	"math"

	"dgs-toolbox/internal/geo"
)

// Krasovsky 1940 ellipsoid, which the conversion is defined on.
const (
	semiMajor    = 6378245.0
	eccentricity = 0.00669342162296594323
)

// InChina reports whether p lies in the rough box the conversion applies to.
// Outside it GCJ-02 equals WGS-84. The box is the conventional one and also
// covers parts of neighbouring countries.
func InChina(p geo.LatLon) bool {
	return p.Lon >= 72.004 && p.Lon <= 137.8347 && p.Lat >= 0.8293 && p.Lat <= 55.8271
}

// FromWGS84 is the GCJ-02 position of a WGS-84 one. Positions outside China
// are returned unchanged.
func FromWGS84(p geo.LatLon) geo.LatLon {
	if !InChina(p) {
		return p
	}
	x, y := p.Lon-105.0, p.Lat-35.0
	dLat := transformLat(x, y)
	dLon := transformLon(x, y)
	radLat := p.Lat / 180.0 * math.Pi
	magic := math.Sin(radLat)
	magic = 1 - eccentricity*magic*magic
	sqrtMagic := math.Sqrt(magic)
	dLat = (dLat * 180.0) / ((semiMajor * (1 - eccentricity)) / (magic * sqrtMagic) * math.Pi)
	dLon = (dLon * 180.0) / (semiMajor / sqrtMagic * math.Cos(radLat) * math.Pi)
	return geo.LatLon{Lat: p.Lat + dLat, Lon: p.Lon + dLon}
}

// ToWGS84 is the WGS-84 position of a GCJ-02 one, found by inverting
// FromWGS84 to well under a millimetre.
func ToWGS84(p geo.LatLon) geo.LatLon {
	if !InChina(p) {
		return p
	}
	w := p
	for range 10 {
		g := FromWGS84(w)
		w.Lat -= g.Lat - p.Lat
		w.Lon -= g.Lon - p.Lon
	}
	return w
}

func transformLat(x, y float64) float64 {
	r := -100.0 + 2.0*x + 3.0*y + 0.2*y*y + 0.1*x*y + 0.2*math.Sqrt(math.Abs(x))
	r += (20.0*math.Sin(6.0*x*math.Pi) + 20.0*math.Sin(2.0*x*math.Pi)) * 2.0 / 3.0
	r += (20.0*math.Sin(y*math.Pi) + 40.0*math.Sin(y/3.0*math.Pi)) * 2.0 / 3.0
	r += (160.0*math.Sin(y/12.0*math.Pi) + 320*math.Sin(y*math.Pi/30.0)) * 2.0 / 3.0
	return r
}

func transformLon(x, y float64) float64 {
	r := 300.0 + x + 2.0*y + 0.1*x*x + 0.1*x*y + 0.1*math.Sqrt(math.Abs(x))
	r += (20.0*math.Sin(6.0*x*math.Pi) + 20.0*math.Sin(2.0*x*math.Pi)) * 2.0 / 3.0
	r += (20.0*math.Sin(x*math.Pi) + 40.0*math.Sin(x/3.0*math.Pi)) * 2.0 / 3.0
	r += (150.0*math.Sin(x/12.0*math.Pi) + 300.0*math.Sin(x/30.0*math.Pi)) * 2.0 / 3.0
	return r
}
